panic while uninstalling part of Elastic Agent

```
panic: sync: negative WaitGroup counter
goroutine 55 [running]:
sync.(*WaitGroup).Add(0xc0002fa930?, 0xc0000ddd40?)
sync/waitgroup.go:80 +0xda\rsync.(*WaitGroup).Done(...)
sync/waitgroup.go:105
[github.com/elastic/elastic-agent/internal/pkg/agent/install.applyDynamics.func1](https://github.com/elastic/elastic-agent/internal/pkg/agent/install.applyDynamics.func1)({0xc0009494d0?, 0xc000524bb0?, 0x6?})
[github.com/elastic/elastic-agent/internal/pkg/agent/install/uninstall.go:240](https://github.com/elastic/elastic-agent/internal/pkg/agent/install/uninstall.go:240) +0x4c
[github.com/elastic/elastic-agent/internal/pkg/composable.(*controller).Run.func1()](https://github.com/elastic/elastic-agent/internal/pkg/composable.(*controller).Run.func1())
[github.com/elastic/elastic-agent/internal/pkg/composable/controller.go:175](https://github.com/elastic/elastic-agent/internal/pkg/composable/controller.go:175) +0x1df
created by [github.com/elastic/elastic-agent/internal/pkg/composable.(*controller).Run](https://github.com/elastic/elastic-agent/internal/pkg/composable.(*controller).Run)
[github.com/elastic/elastic-agent/internal/pkg/composable/controller.go:128](https://github.com/elastic/elastic-agent/internal/pkg/composable/controller.go:128) +0x71b"

```

./elastic-agent install -v -n --url=[xxxx.fleet.us-east-1.aws.found.io:443](https://xxxx.fleet.us-east-1.aws.found.io/) --enrollment-token=xxxxxxxx

--- a/CHANGELOG.next.asciidoc
+++ b/CHANGELOG.next.asciidoc
@@ -112,7 +112,7 @@
 - Allow the / char in variable names in eql and transpiler. {issue}715[715] {pull}718[718]
 - Fix data duplication for standalone agent on Kubernetes using the default manifest {issue-beats}31512[31512] {pull}742[742]
 - Agent updates will clean up unneeded artifacts. {issue}693[693] {issue}694[694] {pull}752[752]
-- Fix a panic cause by a race condition when installing the Elastic Agent. {issues}806[806]
+- Fix a panic caused by a race condition when installing the Elastic Agent. {issues}806[806]
 
 ==== New features
 

--- a/CHANGELOG.next.asciidoc
+++ b/CHANGELOG.next.asciidoc
@@ -112,6 +112,7 @@
 - Allow the / char in variable names in eql and transpiler. {issue}715[715] {pull}718[718]
 - Fix data duplication for standalone agent on Kubernetes using the default manifest {issue-beats}31512[31512] {pull}742[742]
 - Agent updates will clean up unneeded artifacts. {issue}693[693] {issue}694[694] {pull}752[752]
+- Fix a panic cause by a race condition when installing the Elastic Agent. {issues}806[806]
 
 ==== New features
 

--- a/internal/pkg/agent/install/uninstall.go
+++ b/internal/pkg/agent/install/uninstall.go
@@ -235,10 +235,11 @@ func applyDynamics(ctx context.Context, log *logger.Logger, cfg *config.Config) 
 	if ok {
 		varsArray := make([]*transpiler.Vars, 0)
 
-		// Gie some time for the providers to replaces the variables
+		// Give some time for the providers to replace the variables
 		const timeout = 15 * time.Second
 		var doOnce sync.Once
 		ctx, cancel := context.WithTimeout(ctx, timeout)
+		defer cancel()
 
 		// The composable system will continuously run, we are only interested in the first run on of the
 		// renderer to collect the variables we should stop the execution.

--- a/internal/pkg/agent/install/uninstall.go
+++ b/internal/pkg/agent/install/uninstall.go
@@ -235,9 +235,15 @@ func applyDynamics(ctx context.Context, log *logger.Logger, cfg *config.Config) 
 		varsArray := make([]*transpiler.Vars, 0)
 		var wg sync.WaitGroup
 		wg.Add(1)
+
+		ctx, cancel := context.WithCancel(ctx)
+
+		// The composable system will continuously run, we are only interested in the first run on of the
+		// renderer to collect the variables we should stop the execution.
 		varsCallback := func(vv []*transpiler.Vars) {
 			varsArray = vv
 			wg.Done()
+			cancel()
 		}
 
 		ctrl, err := composable.New(log, cfg)

--- a/internal/pkg/agent/install/uninstall.go
+++ b/internal/pkg/agent/install/uninstall.go
@@ -13,6 +13,7 @@ import (
 	"runtime"
 	"strings"
 	"sync"
+	"time"
 
 	"github.com/kardianos/service"
 
@@ -233,17 +234,19 @@ func applyDynamics(ctx context.Context, log *logger.Logger, cfg *config.Config) 
 	inputs, ok := transpiler.Lookup(ast, "inputs")
 	if ok {
 		varsArray := make([]*transpiler.Vars, 0)
-		var wg sync.WaitGroup
-		wg.Add(1)
 
-		ctx, cancel := context.WithCancel(ctx)
+		// Gie some time for the providers to replaces the variables
+		const timeout = 15 * time.Second
+		var doOnce sync.Once
+		ctx, cancel := context.WithTimeout(ctx, timeout)
 
 		// The composable system will continuously run, we are only interested in the first run on of the
 		// renderer to collect the variables we should stop the execution.
 		varsCallback := func(vv []*transpiler.Vars) {
-			varsArray = vv
-			wg.Done()
-			cancel()
+			doOnce.Do(func() {
+				varsArray = vv
+				cancel()
+			})
 		}
 
 		ctrl, err := composable.New(log, cfg)
@@ -251,7 +254,14 @@ func applyDynamics(ctx context.Context, log *logger.Logger, cfg *config.Config) 
 			return nil, err
 		}
 		_ = ctrl.Run(ctx, varsCallback)
-		wg.Wait()
+
+		// Wait for the first callback to retrieve the variables from the providers.
+		<-ctx.Done()
+
+		// Bail out if callback was not executed in time.
+		if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+			return nil, errors.New("failed to get transpiler vars", err)
+		}
 
 		renderedInputs, err := transpiler.RenderInputs(inputs, varsArray)
 		if err != nil {

